### PR TITLE
CSHARP-3140: PartialEvaluator should not evaluate unnecessary clauses for AndAlso, Conditional and OrElse

### PR DESCRIPTION
**Pull request by rstam**

### CODE REVIEW

**Review by JamesKovacs — APPROVED** [review]
LGTM

File: src/MongoDB.Driver/Linq/Linq3Implementation/Misc/PartialEvaluator.cs [hunk 1]
```diff
@@ -71,6 +71,61 @@ public override Expression Visit(Expression expression)
...
+            protected override Expression VisitConditional(ConditionalExpression node)
+            {
+                var test = base.Visit(node.Test);
+                if (test is ConstantExpression constantTestExpression)
+                {
+                    var value = (bool)constantTestExpression.Value;
+                    return value ? Visit(node.IfTrue) : Visit(node.IfFalse);
+                }
+
+                return node.Update(test, Visit(node.IfTrue), Visit(node.IfFalse));
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Misc/PartialEvaluator.cs:126]: Minor optimization that prevents double visitation of test.

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp3140Tests.cs [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+            var currentUser = new User { Id = 1, Factory = null };
+            var queryable = collection.AsQueryable()
+                .Where(x => x.FactoryId != 0 && currentUser.Factory != null);
+
+            var stages = Translate(collection, queryable);
+            AssertStages(stages, "{ $match : { _id : { $type : -1 } } }");
+        }
+
+        [Fact]
+        public void AndAlso_with_neither_clause_a_constant_should_work()
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp3140Tests.cs:72]: New test.

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp3140Tests.cs [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+            var currentUser = new User { Id = 1, Factory = new Factory { Id = 1 } };
+            var queryable = collection.AsQueryable()
+                .Where(x => x.FactoryId == (currentUser.Factory == null ? 0 : currentUser.Factory.Id));
+
+            var stages = Translate(collection, queryable);
+            AssertStages(stages, "{ $match : { FactoryId : 1 } }");
+        }
+
+        [Fact]
+        public void Conditional_with_test_that_is_not_a_constant_should_work()
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp3140Tests.cs:108]: New test.

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp3140Tests.cs [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+            var currentUser = new User { Id = 1, Factory = null };
+            var queryable = collection.AsQueryable()
+                .Where(x => x.FactoryId != 0 || currentUser.Factory == null);
+
+            var stages = Translate(collection, queryable);
+            AssertStages(stages, "{ $match : { } }");
+        }
+
+        [Fact]
+        public void OrElse_with_neither_clause_a_constant_should_work()
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp3140Tests.cs:168]: New test.

**Review by JamesKovacs — CHANGES_REQUESTED** [review]
Minor changes requested.

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp3140Tests.cs [hunk 5]
```diff
@@ -68,6 +68,18 @@ public void AndAlso_with_second_clause_that_evaluates_to_false_should_simplify_t
             AssertStages(stages, "{ $match : { _id : { $type : -1 } } }");
         }
 
+        [Fact]
+        public void AndAlso_with_neither_clause_a_constant_should_work()
+        {
+            var collection = GetCollection<Order>();
+            var currentUser = new User { Id = 1, Factory = null };
```
[the remark this answers is not in the document]
Comment (rstam) [reply tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp3140Tests.cs]: Done

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp3140Tests.cs [hunk 6]
```diff
@@ -92,6 +104,18 @@ public void Conditional_with_test_that_evaluates_to_false_should_simplify_to_if_
             AssertStages(stages, "{ $match : { FactoryId : 1 } }");
         }
 
+        [Fact]
+        public void Conditional_with_test_that_is_not_a_constant_should_work()
+        {
+            var collection = GetCollection<Order>();
+            var currentUser = new User { Id = 1, Factory = new Factory { Id = 1 } };
```
[the remark this answers is not in the document]
Comment (rstam) [reply tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp3140Tests.cs]: Done

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp3140Tests.cs [hunk 7]
```diff
@@ -140,6 +164,18 @@ public void OrElse_with_second_clause_that_evaluates_to_true_should_simplify_to_
             AssertStages(stages, "{ $match : { } }");
         }
 
+        [Fact]
+        public void OrElse_with_neither_clause_a_constant_should_work()
+        {
+            var collection = GetCollection<Order>();
+            var currentUser = new User { Id = 1, Factory = null };
```
[the remark this answers is not in the document]
Comment (rstam) [reply tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp3140Tests.cs]: Done

**Review by JamesKovacs — APPROVED** [review]
LGTM

**Pull request merged**